chiron/multistate.py

```diff
@@ -0,0 +1,968 @@
+    """
+    Base class for samplers that sample multiple thermodynamic states using
+    one or more replicas.
+
+    This base class provides a general simulation facility for multistate from multiple
+    thermodynamic states, allowing any set of thermodynamic states to be specified.
+    If instantiated on its own, the thermodynamic state indices associated with each
+    state are specified and replica mixing does not change any thermodynamic states,
+    meaning that each replica remains in its original thermodynamic state.
+
+    Parameters
+    ----------
+    mcmc_moves : MCMCMove or list of MCMCMove, optional
+        The MCMCMove used to propagate the thermodynamic states. If a list of MCMCMoves,
+        they will be assigned to the correspondent thermodynamic state on
+        creation. If None is provided, Langevin dynamics with 2fm timestep, 5.0/ps collision rate,
+        and 500 steps per iteration will be used.
+    number_of_iterations : int or infinity, optional, default: 1
+        The number of iterations to perform.
+
+    locality : int > 0, optional, default None
+        If None, the energies at all states will be computed for every replica each iteration.
+        If int > 0, energies will only be computed for states ``range(max(0, state-locality), min(n_states, state+locality))``.
+
+    Attributes
+    ----------
+    n_replicas
+    n_states
+    iteration
+    mcmc_moves
+    sampler_states
+    metadata
+    is_completed
+    """
+
+    def __init__(self, mcmc_moves=None, number_of_iterations=1, locality=None):
+        # These will be set on initialization. See function
+        # create() for explanation of single variables.
+        self._thermodynamic_states = None
+        self._unsampled_states = None
+        self._sampler_states = None
+        self._replica_thermodynamic_states = None
+        self._iteration = None
+        self._energy_thermodynamic_states = None
+        self._neighborhoods = None
+        self._energy_unsampled_states = None
+        self._n_accepted_matrix = None
+        self._n_proposed_matrix = None
+        self._reporter = None
+        self._metadata = None
+        self._timing_data = dict()
+
+        # Handling default propagator.
+        if mcmc_moves is None:
+            from .mcmc import LangevinDynamicsMove
+
+            # This will be converted to a list in create().
+            self._mcmc_moves = LangevinDynamicsMove(
+                timestep=2.0 * unit.femtosecond,
+                collision_rate=5.0 / unit.picosecond,
+                n_steps=500,
+            )
+        else:
+            self._mcmc_moves = copy.deepcopy(mcmc_moves)
+
+        # Store constructor parameters. Everything is marked for internal
+        # usage because any change to these attribute implies a change
+        # in the storage file as well. Use properties for checks.
+        self.number_of_iterations = number_of_iterations
+        self._last_mbar_f_k = None
```

Comment:
We might want to encapsulate all the online/offline free energy estimate behavior in a separate object that tracks its own statistics to have fewer things at the "flat" level:
```python
self.online_estimator = None
self.offline_estimator = MBAREstimator()
```

Reply:
I will address this in a separate PR with the reporter class

Reply:
This has been resolved in PR #15